src/randomizer.py

```diff
@@ -865,6 +865,110 @@ def minus1(node: ca.Node) -> ca.BinaryOp:
 
     return True
 
+def perm_struct_ref(
+    fn: ca.FuncDef, ast: ca.FileAST, indices: Indices, region: Region, random: Random
+) -> bool:
+    """Permute struct references: (a + b)->c, and (*(a + b)).c, a[b].c, (&a[b])->c"""
+    cands: List[ca.StructRef] = []
+    class Visitor(ca.NodeVisitor):
+        def visit_StructRef(self, node: ca.StructRef) -> None:
+            if region.contains_node(node):
+                cands.append(node)
+    Visitor().visit(fn.body)
+    if not cands:
+        return False
+
+    # TODO: This might need to handle nested BinaryOps
+    def choose_side(left: ca.Node, right: ca.BinaryOp) -> ca.BinaryOp:
```

Comment:
"randomize_associative_binop"? we should really have this as a standalone randomization pass, that transforms
a + (b ± c) <-> (a + b) ± c
a - (b ± c) <-> (a - b) ∓ c
but of course it doesn't hurt to do it here specifically as well.

This code as well some other seems to assume that + and - are the only BinaryOps that exist. Now, this is *that* big of a problem; if it does something weird with say multiplication we'll just get code that doesn't match very well and move on. But it might be worth fixing for sanity and to improve performance.

Reply:
This is just a hack to properly convert array references, I _think_ it handles other binops on either side, since it uses the right side's operator and doesn't touch the left. I'll have to test it more.